internal/core/software_renderer.rs

```diff
@@ -248,19 +248,67 @@ impl PhysicalRegion {
 
     /// Returns an iterator over the rectangles in this region.
     /// Each rectangle is represented by its position and its size.
+    /// They do not overlap.
     pub fn iter(
         &self,
     ) -> impl Iterator<Item = (crate::api::PhysicalPosition, crate::api::PhysicalSize)> + '_ {
-        self.iter_box().map(|r| {
-            let r = r.to_rect();
-            (
-                crate::api::PhysicalPosition { x: r.origin.x as _, y: r.origin.y as _ },
-                crate::api::PhysicalSize { width: r.width() as _, height: r.height() as _ },
-            )
+        let mut line_ranges = Vec::<core::ops::Range<i16>>::new();
+        let mut begin_line = 0;
+        let mut end_line = 0;
+        line_ranges.reverse();
+        core::iter::from_fn(move || loop {
+            match line_ranges.pop() {
+                Some(r) => {
+                    return Some((
+                        crate::api::PhysicalPosition { x: r.start as _, y: begin_line as _ },
+                        crate::api::PhysicalSize {
+                            width: r.len() as _,
+                            height: (end_line - begin_line) as _,
+                        },
+                    ));
+                }
+                None => {
+                    begin_line = end_line;
+                    end_line = region_line_ranges(self, begin_line, &mut line_ranges)?;
```

Comment:
Took me a while to see the `?` at the end here that makes it a loop that terminates :)